### PR TITLE
Implement common library-init functions

### PR DESCRIPTION
**Pull request by ColdenCullen**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/common.c [hunk 1]
```diff
@@ -352,6 +353,198 @@ void aws_secure_zero(void *pBuf, size_t bufsize) {
...
+        aws_register_error_info(&s_list);
+        aws_register_log_subject_info_list(&s_common_log_subject_list);
+    }
+}
+
+void aws_common_library_clean_up(void) {
+    if (s_common_library_initialized) {
+        s_common_library_initialized = false;
+        aws_unregister_error_info(&s_list);
+        aws_unregister_log_subject_info_list(&s_common_log_subject_list);
```
Comment (graebm) [inline source/common.c]: Just curious, why bother unregistering these? We never did so before

Comment (ColdenCullen) [reply source/common.c]: Step 1: User hand-loads DLL
Step 2: User inits library. aws-c-common now has pointers to strings in DLL static memory
Step 3: User hand-unloads DLL
Step 4: User attempts to print error info. BANG DING OW

File: include/aws/common/common.h [hunk 2]
```diff
@@ -32,4 +32,24 @@
...
+void aws_common_library_init(struct aws_allocator *allocator);
+
+/**
+ * Shuts down the internal datastructures used by aws-c-common.
+ */
+AWS_COMMON_API
+void aws_common_library_clean_up(void);
+
+AWS_COMMON_API
+void aws_common_fatal_assert_library_initialized(void);
```
Comment (graebm) [inline include/aws/common/common.h]: Should probably call this from somewhere. Maybe from `get_error_by_code()`?

File: source/common.c [hunk 3]
```diff
@@ -352,6 +353,210 @@ void aws_secure_zero(void *pBuf, size_t bufsize) {
...
+    }
+}
+
+void aws_common_fatal_assert_library_initialized(void) {
+    if (!s_common_library_initialized) {
+        AWS_LOGF_FATAL(
+            AWS_LS_COMMON_GENERAL,
```
Comment (graebm) [inline source/common.c]: Can't log if the library was never initialized 😭
`fprintf(stderr,` ?

File: source/common.c [hunk 4]
```diff
@@ -352,6 +353,210 @@ void aws_secure_zero(void *pBuf, size_t bufsize) {
...
+
+    if (!s_common_library_initialized) {
+        s_common_library_initialized = true;
+        aws_register_error_info(&s_list);
+        aws_register_log_subject_info_list(&s_common_log_subject_list);
+    }
+}
+
+void aws_common_library_clean_up(void) {
+    if (s_common_library_initialized) {
```
Comment (danielsn) [inline source/common.c]: Should it be an error (or at least a warning) to call clean_up on a library that was never init?

File: source/error.c [hunk 5]
```diff
@@ -150,9 +150,8 @@ void aws_register_error_info(const struct aws_error_info_list *error_info) {
     AWS_ASSERT(error_info->error_list);
     AWS_ASSERT(error_info->count);
 
-    int min_range = error_info->error_list[0].error_code;
-
-    int slot_index = min_range >> SLOT_DIV_SHIFT;
+    const int min_range = error_info->error_list[0].error_code;
+    const int slot_index = min_range >> SLOT_DIV_SHIFT;
```
Comment (danielsn) [inline source/error.c]: Any reasonable compiler will do strength reduction if you do division by a constant power of 2. Would be easier to read.

File: source/common.c [hunk 6]
```diff
@@ -352,6 +353,210 @@ void aws_secure_zero(void *pBuf, size_t bufsize) {
 #endif     // #else not windows
 }
 
+#define AWS_DEFINE_ERROR_INFO_COMMON(C, ES) [(C)-0x0000] = AWS_DEFINE_ERROR_INFO(C, ES, "aws-c-common")
+/* clang-format off */
+static struct aws_error_info errors[] = {
```
Comment (danielsn) [inline source/common.c]: Should there be a way to internationalize this?

File: source/error.c [hunk 7]
```diff
@@ -150,9 +150,8 @@ void aws_register_error_info(const struct aws_error_info_list *error_info) {
...
 
     AWS_ASSERT(slot_index < AWS_MAX_ERROR_SLOTS && slot_index >= 0);
```
Comment (danielsn) [inline source/error.c]: If this is in prod mode, we'll return random data here. Should this either be a fatal assert, or a check that slot_index is in range, followed by returning a "this was a bad slot" error message?

File: source/error.c [hunk 8]
```diff
@@ -182,172 +181,24 @@ void aws_register_error_info(const struct aws_error_info_list *error_info) {
...
-        aws_register_error_info(&s_list);
+void aws_unregister_error_info(const struct aws_error_info_list *error_info) {
+    AWS_ASSERT(error_info);
+    AWS_ASSERT(error_info->error_list);
+    AWS_ASSERT(error_info->count);
+
+    const int min_range = error_info->error_list[0].error_code;
+    const int slot_index = min_range >> SLOT_DIV_SHIFT;
+
+    AWS_ASSERT(slot_index < AWS_MAX_ERROR_SLOTS && slot_index >= 0);
```
Comment (danielsn) [inline source/error.c]: If `AWS_FATAL_ASSERT()` could take an error message string, these could both be combined into a single `AWS_FATAL_ASSERT`

File: source/logging.c [hunk 9]
```diff
@@ -335,9 +333,8 @@ void aws_register_log_subject_info_list(struct aws_log_subject_info_list *log_su
     AWS_ASSERT(log_subject_list->subject_list);
     AWS_ASSERT(log_subject_list->count);
 
-    uint32_t min_range = log_subject_list->subject_list[0].subject_id;
-
-    uint32_t slot_index = min_range >> AWS_LOG_SUBJECT_BIT_SPACE;
+    const uint32_t min_range = log_subject_list->subject_list[0].subject_id;
+    const uint32_t slot_index = min_range >> AWS_LOG_SUBJECT_BIT_SPACE;
 
     AWS_ASSERT(slot_index < AWS_MAX_LOG_SUBJECT_SLOTS);
```
Comment (danielsn) [inline source/logging.c]: Same as before: should this be `AWS_FATAL_ASSERT`?

File: source/logging.c [hunk 10]
```diff
@@ -335,9 +333,8 @@ void aws_register_log_subject_info_list(struct aws_log_subject_info_list *log_su
     AWS_ASSERT(log_subject_list->subject_list);
     AWS_ASSERT(log_subject_list->count);
 
-    uint32_t min_range = log_subject_list->subject_list[0].subject_id;
-
-    uint32_t slot_index = min_range >> AWS_LOG_SUBJECT_BIT_SPACE;
+    const uint32_t min_range = log_subject_list->subject_list[0].subject_id;
+    const uint32_t slot_index = min_range >> AWS_LOG_SUBJECT_BIT_SPACE;
```
Comment (danielsn) [inline source/logging.c]: Same as before - should this be division?

File: source/logging.c [hunk 11]
```diff
@@ -351,22 +348,27 @@ void aws_register_log_subject_info_list(struct aws_log_subject_info_list *log_su
...
-        "task-scheduler",
-        "Subject for task scheduler or task specific logging."),
-};
+void aws_unregister_log_subject_info_list(struct aws_log_subject_info_list *log_subject_list) {
+    /*
+     * We're not so worried about these asserts being removed in an NDEBUG build
+     * - we'll either segfault immediately (for the first two) or for the count
+     * assert, the registration will be ineffective.
+     */
+    AWS_ASSERT(log_subject_list);
```
Comment (danielsn) [inline source/logging.c]: If you expect to segfault immediatly, why not make it a `FATAL_ASSERT` anyway? Its not like this function is on the critical path of a tight loop.

File: source/common.c (hunk 4 above)
Comment (ColdenCullen) [reply source/common.c]: No, the goal is for these functions to be completely idempotent.

File: source/common.c (hunk 6 above)
Comment (ColdenCullen) [reply source/common.c]: Probably, but I believe that to be out of scope for this PR, and likely for GA.

File: source/error.c (hunk 5 above)
Comment (ColdenCullen) [reply source/error.c]: I'm not sure I understand, what would be your proposed implementation?

File: source/error.c (hunk 7 above)
Comment (ColdenCullen) [reply source/error.c]: I can definitely make it a fatal assert. I think this code predates that macro.

Comment (ColdenCullen) [reply source/error.c]: WAit no, it's fatal asserted right below. I can remove this check.

File: source/error.c (hunk 8 above)
Comment (ColdenCullen) [reply source/error.c]: We're logging right below if this condition is bad, so I'm just gonna nuke this.

**Pull request merged**